File: backend/internal/commandstore/feedback.go
package commandstore

import (
	"context"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// Cursor is one device's C2 ack position (device_c2_cursor). The SSE c2cursor producer (Design 23 §4.5 /
// D23.7) diffs the snapshot on applied_seq ONLY; last_poll_at rides the payload for display but never
// triggers a delta (a poll that only bumps last_poll_at must not churn the stream — T9).
type Cursor struct {
	Serial     string     `json:"serial"`
	AppliedSeq int64      `json:"applied_seq"`
	LastPollAt *time.Time `json:"last_poll_at"`
}

// CursorSnapshot reads every device's current C2 cursor — the source for the SSE poll-and-diff producer.
func CursorSnapshot(ctx context.Context, pool *pgxpool.Pool) ([]Cursor, error) {
	rows, err := pool.Query(ctx, `SELECT serial, applied_seq, last_poll_at FROM device_c2_cursor`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []Cursor{}
	for rows.Next() {
		var c Cursor
		if err := rows.Scan(&c.Serial, &c.AppliedSeq, &c.LastPollAt); err != nil {
			return nil, err
		}
		out = append(out, c)
	}
	return out, rows.Err()
}

// RecentCommand is one recently-enqueued command with its per-target applied/pending breakdown — the
// rehydration source the editor seeds from on mount (D23.9), after which the c2cursor SSE carries live
// deltas. The script is DELIBERATELY omitted: the feedback view needs seq/target/note, never the payload,
// and the script can carry URLs/SSIDs (air-gap, T11). "applied" here means the cursor crossed the seq =
// delivered + attempted, NEVER "succeeded" (D23.4 — no device→server result channel).
type RecentCommand struct {
	Seq           int64     `json:"seq"`
	Serial        string    `json:"serial"`
	Note          *string   `json:"note"`
	CreatedAt     time.Time `json:"created_at"`
	OperatorKeyID *int64    `json:"operator_key_id"`
	TargetCount   int       `json:"target_count"`
	AppliedCount  int       `json:"applied_count"`
}

// RecentCommands returns recent command_queue rows (within window, newest first, capped) joined to the
// device cursors for a per-command applied/target breakdown: a single-serial command is applied when that
// device's cursor reached its seq (target=1); a '*' command is the N-of-M across all device cursors
// (target = current roster size; a device onboarded after the seq seeds its cursor past it → counted
// applied, moot, D23.6/D17.7). Read-only — stays reachable under read-only degradation (D23.9).
func RecentCommands(ctx context.Context, pool *pgxpool.Pool, window time.Duration, limit int) ([]RecentCommand, error) {
	const q = `
SELECT cq.seq, cq.serial, cq.note, cq.created_at, cq.operator_key_id,
  CASE WHEN cq.serial = '*' THEN (SELECT count(*)::int FROM devices)
       ELSE 1 END AS target_count,
  CASE WHEN cq.serial = '*'
       THEN (SELECT count(*)::int FROM device_c2_cursor c WHERE c.applied_seq >= cq.seq)
       ELSE (SELECT count(*)::int FROM device_c2_cursor c WHERE c.serial = cq.serial AND c.applied_seq >= cq.seq)
  END AS applied_count
FROM command_queue cq
WHERE cq.created_at >= now() - make_interval(secs => $1)
ORDER BY cq.seq DESC
LIMIT $2`
	rows, err := pool.Query(ctx, q, window.Seconds(), limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []RecentCommand{}
	for rows.Next() {
		var rc RecentCommand
		if err := rows.Scan(&rc.Seq, &rc.Serial, &rc.Note, &rc.CreatedAt, &rc.OperatorKeyID, &rc.TargetCount, &rc.AppliedCount); err != nil {
			return nil, err
		}
		out = append(out, rc)
	}
	return out, rows.Err()
}

File: backend/web/src/lib/berry/feedback.ts
// Berry feedback model (Design 23 §4.5 / D23.4 / D23.9) — the cursor-advance ack, told honestly.
//
// THE load-bearing honesty seam: the device persists the served seq regardless of whether the script
// compiled or ran (cmd.c:371-372), and there is NO device→server result channel — the cursor crossing
// the enqueued seq is the ONLY signal. So "applied" here means DELIVERED + ATTEMPTED, never "succeeded".
// There is deliberately no success/✓ state in this module: a syntactically broken script lands in exactly
// the same "applied" bucket (T4). The only richer outcome is the device log deep-link (the on-device
// `script ok=%d` line, best-effort/async/unstructured — Doc 21).
//
// Initial state seeds from GET /api/berry/commands/recent (rehydration, D23.9 — SSE alone is a forward
// diff with no enqueue history); the c2cursor SSE then carries live applied_seq deltas. applied counts are
// monotonic (cursors only advance), so live data never regresses below the rehydration baseline.

const FLEET = '*'

export interface RecentCommand {
  seq: number
  serial: string
  note: string | null
  created_at: string
  operator_key_id: number | null
  target_count: number
  applied_count: number
}

export interface RecentResponse {
  success: true
  commands: RecentCommand[]
}

export interface C2CursorEvent {
  serial: string
  applied_seq: number
  last_poll_at: string | null
}

/** serial → latest applied_seq, accumulated from c2cursor SSE deltas. */
export type CursorMap = Record<string, number>

export type FeedbackState = 'applied' | 'partial' | 'pending'

export interface FeedbackRow {
  seq: number
  serial: string
  note: string | null
  created_at: string
  state: FeedbackState
  appliedCount: number
  targetCount: number
  /** the honest status line — "delivered + attempted", NEVER "succeeded" (D23.4). */
  label: string
  /** the device to deep-link the log view to (the only place the real outcome may surface); null for '*'. */
  logSerial: string | null
}

/** Validate + narrow a c2cursor SSE payload. Returns null on a malformed frame (never throws). */
export function parseCursorEvent(data: unknown): C2CursorEvent | null {
  if (typeof data !== 'object' || data === null) return null
  const d = data as Record<string, unknown>
  if (typeof d.serial !== 'string' || typeof d.applied_seq !== 'number') return null
  const lp = typeof d.last_poll_at === 'string' ? d.last_poll_at : null
  return { serial: d.serial, applied_seq: d.applied_seq, last_poll_at: lp }
}

/** Fold one cursor delta into the map (cursors only move forward — keep the max, never regress). */
export function applyCursorEvent(cursors: CursorMap, ev: C2CursorEvent): CursorMap {
  const prev = cursors[ev.serial] ?? -1
  if (ev.applied_seq <= prev) return cursors
  return { ...cursors, [ev.serial]: ev.applied_seq }
}

/** How many known device cursors have reached `seq` (the live '*' applied count). */
function appliedAcrossFleet(cursors: CursorMap, seq: number): number {
  let n = 0
  for (const s of Object.keys(cursors)) if (cursors[s] >= seq) n++
  return n
}

/**
 * Resolve one recent command to a feedback row, blending the rehydration baseline with live cursor data.
 * applied counts are monotonic: live data can only raise the count (cursors advance), capped at the target.
 */
export function resolveCommand(cmd: RecentCommand, cursors: CursorMap): FeedbackRow {
  const isFleet = cmd.serial === FLEET
  const targetCount = cmd.target_count

  let appliedCount: number
  if (isFleet) {
    appliedCount = Math.min(targetCount, Math.max(cmd.applied_count, appliedAcrossFleet(cursors, cmd.seq)))
  } else {
    const live = cmd.serial in cursors ? (cursors[cmd.serial] >= cmd.seq ? 1 : 0) : cmd.applied_count
    appliedCount = Math.max(cmd.applied_count, live)
  }

  const state: FeedbackState =
    appliedCount >= targetCount ? 'applied' : appliedCount > 0 ? 'partial' : 'pending'

  return {
    seq: cmd.seq,
    serial: cmd.serial,
    note: cmd.note,
    created_at: cmd.created_at,
    state,
    appliedCount,
    targetCount,
    label: labelFor(state, appliedCount, targetCount, isFleet),
    logSerial: isFleet ? null : cmd.serial,
  }
}

/** The honest label. "delivered + attempted" on every applied/partial row — NEVER a success claim (D23.4):
 *  the word is avoided entirely so neither the operator nor a future grep mistakes it for a result. */
function labelFor(state: FeedbackState, applied: number, target: number, isFleet: boolean): string {
  const tail = 'delivered + attempted (not proof it ran — check the device log)'
  if (state === 'applied') {
    return isFleet ? `${applied} of ${target} applied · ${tail}` : `applied · ${tail}`
  }
  if (state === 'partial') {
    return `${applied} of ${target} applied · ${tail} · rest pending`
  }
  // pending
  return isFleet
    ? `0 of ${target} applied — awaiting device polls (no ETA)`
    : 'pending — awaiting the device’s next poll (no ETA)'
}

/** Build the full feedback list (newest-first preserved from the rehydration order). */
export function buildFeedback(commands: RecentCommand[], cursors: CursorMap): FeedbackRow[] {
  return commands.map((c) => resolveCommand(c, cursors))
}
